store/tikv/lock_resolver.go

```diff
@@ -578,10 +581,88 @@ func (lr *LockResolver) getTxnStatus(bo *Backoffer, txnID uint64, primary []byte
 			status.commitTS = cmdResp.CommitVersion
 			lr.saveResolved(txnID, status)
 		}
+		if cmdResp.UseAsyncCommit {
+			status.secondaries = cmdResp.GetSecondaries()
+		}
+
 		return status, nil
 	}
 }
 
+// resolveLockAsync resolves l assuming it was locked using the async commit protocol. Precondition: status.secondaries is non-empty.
+func (lr *LockResolver) resolveLockAsync(bo *Backoffer, l *Lock, status TxnStatus, lite bool, cleanRegions map[RegionVerID]struct{}) error {
+	// TODO(nrc) incr a counter?
+
+	regions, _, err := lr.store.GetRegionCache().GroupKeysByRegion(bo, status.secondaries, nil)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	var commitTs uint64 = 0
+	var locks []*Lock
+
+regionLoop:
+	for regionID, keys := range regions {
+		getReq := &kvrpcpb.BatchGetRequest{
+			Keys:    keys,
+			Version: l.TxnID,
+		}
+		req := tikvrpc.NewRequest(tikvrpc.CmdBatchGet, getReq)
+		resp, err := lr.store.SendReq(bo, req, regionID, readTimeoutShort)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		regionErr, err := resp.GetRegionError()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if regionErr != nil {
+			err = bo.Backoff(BoRegionMiss, errors.New(regionErr.String()))
+			if err != nil {
+				return errors.Trace(err)
+			}
+			continue
+		}
+		if resp.Resp == nil {
+			return errors.Trace(ErrBodyMissing)
+		}
+		getResp := resp.Resp.(*kvrpcpb.BatchGetResponse)
+
+		for _, kv := range getResp.Pairs {
+			if e := kv.Error; e != nil && e.Locked != nil {
+				lockInfo := e.Locked
+				if lockInfo.LockVersion == l.TxnID {
+					if lockInfo.MinCommitTs > commitTs {
+						commitTs = lockInfo.MinCommitTs
+					}
+					locks = append(locks, NewLock(kv.Error.Locked))
+					continue
+				}
+			}
+			// kv is not locked or locked by another transaction
+			commitTs = 0
+			break regionLoop
+		}
+	}
+
+	status.commitTS = commitTs
+	// Resolve every lock in the transaction.
+	// TODO can do secondary (but not primary) locks in parallel.
+	for _, lock := range locks {
+		err := lr.resolveLock(bo, lock, status, lite, cleanRegions)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
```

Comment:
Sorry, but now I think we may need a new command for this work.

Using ResolveLock has a problem: The first `BatchGet` does not meet the lock. Then, another client prewrites the secondary lock successfully and report success to the user. Here we use `commitTS = 0` to roll back the lock and breaks the atomicity of the transaction.

The new command should return the lock if exists, or write a rollback. It's like `CheckTxnStatus`, but for secondary keys.

Reply:
Maybe we need a `checkKeyStatus` command to make the `check lock` and `resolve lock` atomic for the recovery logic?

Reply:
Sounds like the rollback record for secondaries of async commit transactions need to be protected?